rclcpp/src/rclcpp/context.cpp

```diff
@@ -35,21 +35,91 @@
 
 #include "./logging_mutex.hpp"
 
-/// Mutex to protect initialized contexts.
-static std::mutex g_contexts_mutex;
-/// Weak list of context to be shutdown by the signal handler.
-static std::vector<std::weak_ptr<rclcpp::Context>> g_contexts;
+using rclcpp::Context;
+
+namespace rclcpp
+{
+/// Class to manage vector of weak pointers to all created contexts
+struct WeakContextsWrapper
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(WeakContextsWrapper)
+
+  void
+  add_context(const Context::SharedPtr & context)
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    weak_contexts_.push_back(context);
+  }
+
+  void
+  remove_context(const Context * context)
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    weak_contexts_.erase(
+      std::remove_if(
+        weak_contexts_.begin(),
+        weak_contexts_.end(),
+        [context](const Context::WeakPtr weak_context) {
+          auto locked_context = weak_context.lock();
+          if (!locked_context) {
+            // take advantage and removed expired contexts
+            return true;
+          } else {
```

Comment:
@ivanpauno nit: `else` clause is redundant.

Reply:
Fixed in https://github.com/ros2/rclcpp/pull/1132/commits/2b081fc7b26b74c4a25da7c1b9df8844fcd7702b.